funsor/tensor.py

```diff
@@ -0,0 +1,973 @@
+                    raise ValueError("Size mismatch at {}: {} vs {}"
+                                     .format(name, size, other_size))
+        output = reals(*(size_dict[d] for d in ein_output))
+        super(Einsum, self).__init__(inputs, output)
+        self.equation = equation
+        self.operands = operands
+
+    def __repr__(self):
+        return 'Einsum({}, {})'.format(repr(self.equation), repr(self.operands))
+
+    def __str__(self):
+        return 'Einsum({}, {})'.format(repr(self.equation), str(self.operands))
+
+
+@eager.register(Einsum, str, tuple)
+def eager_einsum(equation, operands):
+    if all(isinstance(x, Tensor) for x in operands):
+        # Make new symbols for inputs of operands.
+        inputs = OrderedDict()
+        for x in operands:
+            inputs.update(x.inputs)
+        symbols = set(equation)
+        get_symbol = iter(map(opt_einsum.get_symbol, itertools.count()))
+        new_symbols = {}
+        for k in inputs:
+            symbol = next(get_symbol)
+            while symbol in symbols:
+                symbol = next(get_symbol)
+            symbols.add(symbol)
+            new_symbols[k] = symbol
+
+        # Manually broadcast using einsum symbols.
+        assert '.' not in equation
+        ins, out = equation.split('->')
+        ins = ins.split(',')
+        ins = [''.join(new_symbols[k] for k in x.inputs) + x_out
+               for x, x_out in zip(operands, ins)]
+        out = ''.join(new_symbols[k] for k in inputs) + out
+        equation = ','.join(ins) + '->' + out
+
+        data = ops.einsum(equation, *[x.data for x in operands])
+        return Tensor(data, inputs)
+
+    return None  # defer to default implementation
+
+
+def numeric_tensordot(x, y, dims):
+    """
+    Wrapper around :func:`torch.tensordot` or :func:`numpy.tensordot`
+    to operate on real-valued Funsors.
+
+    Note this operates only on the ``output`` tensor. To perform sum-product
+    contractions on named dimensions, instead use ``+`` and
+    :class:`~funsor.terms.Reduce`.
+
+    Arguments should satisfy::
+
+        len(x.shape) >= dims
+        len(y.shape) >= dims
+        dims == 0 or x.shape[-dims:] == y.shape[:dims]
+
+    :param Funsor x: A left hand argument.
+    :param Funsor y: A y hand argument.
+    :param int dims: The number of dimension of overlap of output shape.
+    :rtype: Funsor
+    """
+    assert dims >= 0
+    assert len(x.shape) >= dims
+    assert len(y.shape) >= dims
+    assert dims == 0 or x.shape[-dims:] == y.shape[:dims]
+    x_start, x_end = 0, len(x.output.shape)
+    y_start = x_end - dims
+    y_end = y_start + len(y.output.shape)
+    symbols = 'abcdefghijklmnopqrstuvwxyz'
+    equation = '{},{}->{}'.format(symbols[x_start:x_end],
+                                  symbols[y_start:y_end],
+                                  symbols[x_start:y_start] + symbols[x_end:y_end])
+    return Einsum(equation, (x, y))
+
+
+def numeric_stack(parts, dim=0):
```

Comment:
ditto: torch_stack -> numeric_stack

Reply:
ditto, I think simply `stack()` or `tensor_stack()` makes sense now. We have a different `Stack()` for the funsor version.